gui/journal.lua

```diff
@@ -0,0 +1,797 @@
+    local from_last_click_ms = (os.clock() - (self.last_click or 0)) * 1000
+
+    if (from_last_click_ms > 250) then
+        self.clicks_count = 0;
+    end
+
+    return self.clicks_count
+end
+
+function TextEditorView:triggerMultiLeftClick()
+    local clicks_count = self:getMultiLeftClick()
+
+    self.clicks_count = clicks_count + 1
+    if (self.clicks_count >= 4) then
+        self.clicks_count = 1
+    end
+
+    self.last_click = os.clock()
+    return self.clicks_count
+end
+
+function TextEditorView:currentSpacesRange()
+    local ind = self:cursorToIndex(self.cursor.x, self.cursor.y)
+    -- select "word" only from spaces
+    local prev_word_end, _  = self.text
+        :sub(1, ind)
+        :find('[^%s]%s+$')
+    local _, next_word_start = self.text:find('%s[^%s]', ind)
+
+    return {
+        x_from=prev_word_end and self.cursor.x - (ind - prev_word_end) + 1 or 1,
+        x_to=next_word_start and self.cursor.x + next_word_start - ind - 1 or #self.text
+    }
+end
+
+function TextEditorView:currentWordRange()
+    -- select current word
+    local ind = self:cursorToIndex(self.cursor.x, self.cursor.y)
+    local _, prev_word_end = self.text
+        :sub(1, ind-1)
+        :find('.*[%s,."\']')
+    local next_word_start, _  = self.text:find('[%s,."\']', ind)
+
+    return {
+        x_from=prev_word_end and self.cursor.x - (ind - prev_word_end) + 1 or 1,
+        x_to=next_word_start and self.cursor.x + next_word_start - ind - 1 or #self.text
+    }
+end
+
+function TextEditorView:onInput(keys)
+    for _,ignore_key in ipairs(self.ignore_keys) do
+        if keys[ignore_key] then
+            return false
+        end
+    end
+
+    if keys.SELECT then
+        -- handle enter
+        self:insert(NEWLINE)
+        return true
+
+    elseif keys._MOUSE_L then
+        local mouse_x, mouse_y = self:getMousePos()
+        if mouse_x and mouse_y then
+
+            local clicks_count = self:triggerMultiLeftClick()
+            if clicks_count >= 3 then
+                self:setSelection(
+                    1,
+                    self.cursor.y,
+                    #self.lines[self.cursor.y],
+                    self.cursor.y
+                )
+            elseif clicks_count >= 2 then
+                local cursor_char = self:charAtCursor()
+
+                local word_range = (
+                    cursor_char == ' ' or cursor_char == NEWLINE
+                ) and self:currentSpacesRange() or self:currentWordRange()
+
+                self:setSelection(
+                    word_range.x_from,
+                    self.cursor.y,
+                    word_range.x_to,
+                    self.cursor.y
+                )
+            elseif clicks_count == 1 then
+                y = math.min(#self.lines, mouse_y + 1)
+                x = math.min(#self.lines[y], mouse_x + 1)
+                self:setCursor(x, y)
+            end
+
+            return true
+        end
+
+    elseif keys._MOUSE_L_DOWN then
```

Comment:
nice. I was worried that you'd choose one of the older methods for detecting dragging (e.g. scanning `enabler` in the render method). This is the right way to do it in modern code.